### DISCUSSION TITLE
Deferred shading. Position Conversion

### DISCUSSION THREAD

**Discussion by AndreyPrakhov** [pmndrs/postprocessing#618] [Q&A] [answered]
In the document "Redesign of the rendering pipeline", GBuffer assumed the output of the "position". He's not here right now. I guess this was abandoned in favor of converting the pixel position from the depth buffer. I am currently working on a lighting implementation from PointLight and I am trying to use this feature in a shader.

```
float viewZ = getViewZ( gData.depth, cameraParams.x, cameraParams.y);
vec3 vViewPosition = getViewPosition(uv, gData.depth, viewZ,projectionMatrix,projectionMatrixInverse);
```
Of course, nothing happens so easily and the light source begins to wander around the stage )

I have two questions:

- What's wrong with using the getViewPosition function
- Is there a way to add a position calculation to GBuffer using normal means, rather than rewriting part of the code

**Comment by AndreyPrakhov** [comment]
Hmm, the first question is removed, I calculated vViewPosition incorrectly. If anyone needs it, it looks like this:

```
vec3 position = getViewPosition(uv, gData.depth, viewZ,projectionMatrix,projectionMatrixInverse);
vec4 mvPosition = viewMatrix * vec4( position, 1.0 );
vec3 vViewPosition = - mvPosition.xyz;
```
The question of how to add your own buffers to GBuffer remains relevant

**Comment by vanruesc** [comment]
> how to add your own buffers to GBuffer

That should technically be possible by passing a [GBufferConfig](https://github.com/pmndrs/postprocessing/blob/v7/src/utils/GBufferConfig.ts) with additional data into [GeometryPass](https://github.com/pmndrs/postprocessing/blob/b4934fe06d969d2e25f6f92aa848e17d140b942f/src/passes/GeometryPass.ts#L81). Rendering to the custom texture attachment would have to be implemented by modifying one of three's `ShaderChunk` entries. This is currently untested though.

> I calculated vViewPosition incorrectly. If anyone needs it, it looks like this

Thanks for that snippet. The `position` data is currently missing from `GData`, but I'm planning on adding it as an auto-computed field similar to how `GData.luminance` is implemented.

**Comment by vanruesc** [comment, marked as answer]
I've added support for `GData.POSITION` in `postprocessing@7.0.0-alpha.4`. If an Effect specifies `GBuffer.DEPTH` as input and accesses `gData.position`, the field will be populated with the view position.

From what I've read, lighting calculations should either be done in view space or world space. If you need the world position, you can `#include <pp_world_utils_pars_fragment>` and then use `getWorldPosition(gData.position, uv)`.

**Comment by vanruesc** [comment]
Also, the parameters of `getViewZ` and `getViewPosition` have changed:
-  `near` and `far` have been removed from `getViewZ()`
- `projectionMatrix` and `projectionMatrixInverse` have been removed from `getViewPosition()`
  - `viewZ` is now optional via an overload.
